### DISCUSSION TITLE
Why WebAuthnCBORModule and WebAuthnJSONModule have different serializer ?

### DISCUSSION THREAD

**Discussion by dscreve** [webauthn4j/webauthn4j#1063] [Q&A] [answered]
AAGUID Serializer and deserializer are missing from WebAuthnJSONModule
Challenge Serializer and deserializer are missing from WebAuthnCBORModule

Why there's not the same serializers ?

David

**Comment by ynojima** [comment, marked as answer]
The serializer classes in WebAuthn4J are designed to serialize envelope types defined in the WebAuthn specification, FIDO CTAP2 specification, or FIDO MDS3 specification.
Conversely, serializers that are not required for those types are not provided.
You may implement your own (de)serializers.

**Reply by dscreve** [reply]
You mean that Challenge does not need to be serialized if I use CBOR serializer ?

**Reply by ynojima** [reply]
In the WebAuthn specification, `Challenge` type is used only in JSON-serialized types, like PublicKeyCredentialCreationOptionsJSON.
Of course, if you want to define your own envelope type that contains `Challenge` and serialize it using CBOR, you can achieve this by implementing a custom CBOR serializer.

**Comment by dscreve** [comment]
The problem is when serializing CredentialRecordImpl.

If you want to serialize as CBOR, CollectedClientData cannot be serialized/deserialized due to Challenge.

If you want to serialize as JSON, this is AttestationStatementEnvelope that cannot be serialized/deserialized.

I’ve succeeded in serializing to JSON/CBOR but it’s a really painful code that mixes CBOR / JSON serialization.


--
David Scrève - DLTA Studio
Gérant - 06 27 69 58 14
***@***.***
4249534F5553

**Reply by ynojima** [reply]
Since WebAuthn messages are a mix of CBOR and JSON, it's inevitable that the fields in CredentialRecord are a mix of JSON and CBOR as well.
Please refer [this section for serialization CredentialRecord](https://webauthn4j.github.io/webauthn4j/en/#credentialrecord-serialization-and-deserialization).

**Comment by dscreve** [comment]
Tasks for your help !You did a great job !David ScrèveLe 1 déc. 2024 à 03:52, Yoshikazu Nojima ***@***.***> a écrit :﻿
Since WebAuthn messages are a mix of CBOR and JSON, it's inevitable that the fields in CredentialRecord are a mix of JSON and CBOR as well.
Please refer this section for serialization CredentialRecord.
